p2p/gossip/protocol.go

```diff
@@ -18,14 +18,51 @@ const propagateHandleBufferSize = 1000 // number of MessageValidation that we al
 const ProtocolName = "/p2p/1.0/gossip"
 const protocolVer = "0"
 
-type hash uint32
+type hash [12]byte
 
 // fnv.New32 must be used every time to be sure we get consistent results.
 func calcHash(msg []byte, prot string) hash {
-	msghash := fnv.New32() // todo: Add nonce to messages instead
-	msghash.Write(msg)
-	msghash.Write([]byte(prot))
-	return hash(msghash.Sum32())
+	msghash := sha256.Sum256(append(msg, []byte(prot)...))
+	var h [12]byte
+	copy(h[:], msghash[0:12])
+	return hash(h)
+}
+
+type doubleCache struct {
+	size   uint
+	cacheA map[hash]struct{}
+	cacheB map[hash]struct{}
+}
+
+func newDoubleCache(size uint) *doubleCache {
+	return &doubleCache{size, make(map[hash]struct{}, size), make(map[hash]struct{}, size)}
+}
+
+func (a *doubleCache) lookupOrCreate(key hash) bool {
```

Comment:
perhaps this should be called just lookup, the fact that it stores the value shouldn't matter... what do you think?

Reply:
i actually think if this method was only lookup someone could mistakenly try to insert after looking up. though maybe a better name is needed I'll try to come up with it